### PR TITLE
[Merged by Bors] - feat(Algebra/Group/Pointwise/Set/Basic): sInter and sUnion forms of pointwise operations

### PR DESCRIPTION
**Pull request by mans0954**
Provides `sInter` and `sUnion` versions of  pointwise algebraic operations on sets.

Needed in #17029

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Group/Pointwise/Set/Basic.lean [hunk 1]
```diff
@@ -697,6 +737,14 @@ lemma iUnion_smul (s : ι → Set α) (t : Set β) : (⋃ i, s i) • t = ⋃ i,
 lemma smul_iUnion (s : Set α) (t : ι → Set β) : (s • ⋃ i, t i) = ⋃ i, s • t i :=
   image2_iUnion_right _ _ _
 
+@[to_additive]
+lemma sUnion_smul (S : Set (Set α)) (t : Set β) : (⋃₀ S) • t = ⋃ s ∈ S, s • t :=
```
Comment (YaelDillies) [inline Mathlib/Algebra/Group/Pointwise/Set/Basic.lean:741]:
```suggestion
lemma sUnion_smul (S : Set (Set α)) (t : Set β) : ⋃₀ S • t = ⋃ s ∈ S, s • t :=
```

File: Mathlib/Algebra/Group/Pointwise/Set/Basic.lean [hunk 2]
```diff
@@ -590,6 +622,14 @@ theorem iInter_div_subset (s : ι → Set α) (t : Set α) : (⋂ i, s i) / t 
 theorem div_iInter_subset (s : Set α) (t : ι → Set α) : (s / ⋂ i, t i) ⊆ ⋂ i, s / t i :=
   image2_iInter_subset_right _ _ _
 
+@[to_additive]
+theorem sInter_div_subset (S : Set (Set α)) (t : Set α) : (⋂₀ S) / t ⊆ ⋂ s ∈ S, s / t :=
```
Comment (YaelDillies) [inline Mathlib/Algebra/Group/Pointwise/Set/Basic.lean:626]:
```suggestion
theorem sInter_div_subset (S : Set (Set α)) (t : Set α) : ⋂₀ S / t ⊆ ⋂ s ∈ S, s / t :=
```

File: Mathlib/Algebra/Group/Pointwise/Set/Basic.lean [hunk 3]
```diff
@@ -568,6 +592,14 @@ theorem iUnion_div (s : ι → Set α) (t : Set α) : (⋃ i, s i) / t = ⋃ i,
 theorem div_iUnion (s : Set α) (t : ι → Set α) : (s / ⋃ i, t i) = ⋃ i, s / t i :=
   image2_iUnion_right _ _ _
 
+@[to_additive]
+theorem sUnion_div (S : Set (Set α)) (t : Set α) : (⋃₀ S) / t = ⋃ s ∈ S, s / t :=
```
Comment (YaelDillies) [inline Mathlib/Algebra/Group/Pointwise/Set/Basic.lean:596]:
```suggestion
theorem sUnion_div (S : Set (Set α)) (t : Set α) : ⋃₀ S / t = ⋃ s ∈ S, s / t :=
```

File: Mathlib/Algebra/Group/Pointwise/Set/Basic.lean [hunk 4]
```diff
@@ -380,6 +388,14 @@ theorem iUnion_mul (s : ι → Set α) (t : Set α) : (⋃ i, s i) * t = ⋃ i,
 theorem mul_iUnion (s : Set α) (t : ι → Set α) : (s * ⋃ i, t i) = ⋃ i, s * t i :=
   image2_iUnion_right _ _ _
 
+@[to_additive]
+theorem sUnion_mul (S : Set (Set α)) (t : Set α) : (⋃₀ S) * t = ⋃ s ∈ S, s * t :=
+  image2_sUnion_left _ _ _
+
+@[to_additive]
+theorem mul_sUnion (s : Set α) (T : Set (Set α)) : (s * ⋃₀ T) = ⋃ t ∈ T, s * t :=
```
Comment (YaelDillies) [inline Mathlib/Algebra/Group/Pointwise/Set/Basic.lean:396]:
```suggestion
theorem mul_sUnion (s : Set α) (T : Set (Set α)) : s * ⋃₀ T = ⋃ t ∈ T, s * t :=
```

File: Mathlib/Algebra/Group/Pointwise/Set/Basic.lean [hunk 5]
```diff
@@ -380,6 +388,14 @@ theorem iUnion_mul (s : ι → Set α) (t : Set α) : (⋃ i, s i) * t = ⋃ i,
 theorem mul_iUnion (s : Set α) (t : ι → Set α) : (s * ⋃ i, t i) = ⋃ i, s * t i :=
   image2_iUnion_right _ _ _
 
+@[to_additive]
+theorem sUnion_mul (S : Set (Set α)) (t : Set α) : (⋃₀ S) * t = ⋃ s ∈ S, s * t :=
```
Comment (YaelDillies) [inline Mathlib/Algebra/Group/Pointwise/Set/Basic.lean:392]:
```suggestion
theorem sUnion_mul (S : Set (Set α)) (t : Set α) : ⋃₀ S * t = ⋃ s ∈ S, s * t :=
```

File: Mathlib/Algebra/Group/Pointwise/Set/Basic.lean [hunk 6]
```diff
@@ -868,6 +933,12 @@ lemma iUnion_vsub (s : ι → Set β) (t : Set β) : (⋃ i, s i) -ᵥ t = ⋃ i
 lemma vsub_iUnion (s : Set β) (t : ι → Set β) : (s -ᵥ ⋃ i, t i) = ⋃ i, s -ᵥ t i :=
   image2_iUnion_right _ _ _
 
+lemma sUnion_vsub (S : Set (Set β)) (t : Set β) : (⋃₀ S) -ᵥ t = ⋃ s ∈ S, s -ᵥ t :=
```
Comment (YaelDillies) [inline Mathlib/Algebra/Group/Pointwise/Set/Basic.lean:936]:
```suggestion
lemma sUnion_vsub (S : Set (Set β)) (t : Set β) : ⋃₀ S -ᵥ t = ⋃ s ∈ S, s -ᵥ t :=
```

File: Mathlib/Algebra/Group/Pointwise/Set/Basic.lean [hunk 7]
```diff
@@ -880,6 +951,12 @@ lemma iInter_vsub_subset (s : ι → Set β) (t : Set β) : (⋂ i, s i) -ᵥ t
 lemma vsub_iInter_subset (s : Set β) (t : ι → Set β) : (s -ᵥ ⋂ i, t i) ⊆ ⋂ i, s -ᵥ t i :=
   image2_iInter_subset_right _ _ _
 
+lemma sInter_vsub_subset (S : Set (Set β)) (t : Set β) : (⋂₀ S ) -ᵥ t ⊆ ⋂ s ∈ S, s -ᵥ t :=
```
Comment (YaelDillies) [inline Mathlib/Algebra/Group/Pointwise/Set/Basic.lean:954]:
```suggestion
lemma sInter_vsub_subset (S : Set (Set β)) (t : Set β) : ⋂₀ S -ᵥ t ⊆ ⋂ s ∈ S, s -ᵥ t :=
```

File: Mathlib/Algebra/Group/Pointwise/Set/Basic.lean [hunk 8]
```diff
@@ -868,6 +933,12 @@ lemma iUnion_vsub (s : ι → Set β) (t : Set β) : (⋃ i, s i) -ᵥ t = ⋃ i
 lemma vsub_iUnion (s : Set β) (t : ι → Set β) : (s -ᵥ ⋃ i, t i) = ⋃ i, s -ᵥ t i :=
   image2_iUnion_right _ _ _
 
+lemma sUnion_vsub (S : Set (Set β)) (t : Set β) : (⋃₀ S) -ᵥ t = ⋃ s ∈ S, s -ᵥ t :=
+  image2_sUnion_left _ _ _
+
+lemma vsub_sUnion (s : Set β) (T : Set (Set β)) : (s -ᵥ ⋃₀ T) = ⋃ t ∈ T, s -ᵥ t :=
```
Comment (YaelDillies) [inline Mathlib/Algebra/Group/Pointwise/Set/Basic.lean:939]:
```suggestion
lemma vsub_sUnion (s : Set β) (T : Set (Set β)) : s -ᵥ ⋃₀ T = ⋃ t ∈ T, s -ᵥ t :=
```

File: Mathlib/Algebra/Group/Pointwise/Set/Basic.lean [hunk 9]
```diff
@@ -880,6 +951,12 @@ lemma iInter_vsub_subset (s : ι → Set β) (t : Set β) : (⋂ i, s i) -ᵥ t
 lemma vsub_iInter_subset (s : Set β) (t : ι → Set β) : (s -ᵥ ⋂ i, t i) ⊆ ⋂ i, s -ᵥ t i :=
   image2_iInter_subset_right _ _ _
 
+lemma sInter_vsub_subset (S : Set (Set β)) (t : Set β) : (⋂₀ S ) -ᵥ t ⊆ ⋂ s ∈ S, s -ᵥ t :=
+  image2_sInter_subset_left _ _ _
+
+lemma vsub_sInter_subset (s : Set β) (T : Set (Set β)) : (s -ᵥ ⋂₀ T) ⊆ ⋂ t ∈ T, s -ᵥ t :=
```
Comment (YaelDillies) [inline Mathlib/Algebra/Group/Pointwise/Set/Basic.lean:957]:
```suggestion
lemma vsub_sInter_subset (s : Set β) (T : Set (Set β)) : s -ᵥ ⋂₀ T ⊆ ⋂ t ∈ T, s -ᵥ t :=
```

File: Mathlib/Data/Set/Lattice.lean [hunk 10]
```diff
@@ -1575,6 +1583,15 @@ theorem image2_iUnion_right (s : Set α) (t : ι → Set β) :
...
   simp only [← image_prod, prod_iUnion, image_iUnion]
 
+theorem image2_sUnion_left (S : Set (Set α)) (t : Set β) :
+    image2 f (⋃₀ S) t = ⋃ s ∈ S, image2 f s t := by
+  aesop
+
+theorem image2_sUnion_right (s : Set α) (T : Set (Set β)) :
+    image2 f s (⋃₀ T) = ⋃ t ∈ T, image2 f s t := by
+  aesop
+
```
Comment (YaelDillies) [inline Mathlib/Data/Set/Lattice.lean:1593]:
```suggestion
```

File: Mathlib/Algebra/Group/Pointwise/Set/Basic.lean [hunk 11]
```diff
@@ -779,6 +840,10 @@ lemma smul_set_iUnion₂ (a : α) (s : ∀ i, κ i → Set β) :
 lemma smul_set_iInter_subset (a : α) (t : ι → Set β) : a • ⋂ i, t i ⊆ ⋂ i, a • t i :=
   image_iInter_subset _ _
 
+@[to_additive]
+lemma smul_set_sInter_subset (a : α) (S : Set (Set β)) :
+    a • ⋂₀ S ⊆ ⋂ s ∈ S, a • s := image_sInter_subset S (fun x => a • x)
```
Comment (YaelDillies) [inline Mathlib/Algebra/Group/Pointwise/Set/Basic.lean:845]:
```suggestion
    a • ⋂₀ S ⊆ ⋂ s ∈ S, a • s := image_sInter_subset ..
```
Pretty sure this works

File: Mathlib/Algebra/Group/Pointwise/Set/Basic.lean [hunk 12]
```diff
@@ -771,6 +827,11 @@ lemma smul_set_inter_subset : a • (t₁ ∩ t₂) ⊆ a • t₁ ∩ a • t
 lemma smul_set_iUnion (a : α) (s : ι → Set β) : a • ⋃ i, s i = ⋃ i, a • s i :=
   image_iUnion
 
+@[to_additive]
+lemma smul_set_sUnion (a : α) (S : Set (Set β)) : a • ⋃₀ S = ⋃ s ∈ S, a • s := by
+  convert image_sUnion
+  exact Eq.symm (sUnion_image (image fun x ↦ a • x) S)
```
Comment (YaelDillies) [inline Mathlib/Algebra/Group/Pointwise/Set/Basic.lean:833]:
```suggestion
  rw [sUnion_eq_biUnion, smul_set_iUnion₂]
```

File: Mathlib/Algebra/Group/Pointwise/Set/Basic.lean [hunk 13]
```diff
@@ -868,6 +933,12 @@ lemma iUnion_vsub (s : ι → Set β) (t : Set β) : (⋃ i, s i) -ᵥ t = ⋃ i
 lemma vsub_iUnion (s : Set β) (t : ι → Set β) : (s -ᵥ ⋃ i, t i) = ⋃ i, s -ᵥ t i :=
   image2_iUnion_right _ _ _
 
+lemma sUnion_vsub (S : Set (Set β)) (t : Set β) : ⋃₀ S -ᵥ t = ⋃ s ∈ S, s -ᵥ t :=
+  image2_sUnion_left _ _ _
```
Comment (YaelDillies) [inline Mathlib/Algebra/Group/Pointwise/Set/Basic.lean:937]:
```suggestion
  image2_sUnion_left ..
```
Same everywhere else

File: Mathlib/Data/Set/Lattice.lean [hunk 14]
```diff
@@ -1596,6 +1612,16 @@ theorem image2_iInter_subset_right (s : Set α) (t : ι → Set β) :
   simp_rw [image2_subset_iff, mem_iInter]
   exact fun x hx y hy i => mem_image2_of_mem hx (hy _)
 
+theorem image2_sInter_subset_left (S : Set (Set α)) (t : Set β) :
+    image2 f (⋂₀ S) t ⊆ ⋂ s ∈ S, image2 f s t := by
+  simp_rw [image2_subset_iff, mem_sInter]
+  exact fun x hx y hy i => by aesop
```
Comment (YaelDillies) [inline Mathlib/Data/Set/Lattice.lean:1618]:
```suggestion
    image2 f (⋂₀ S) t ⊆ ⋂ s ∈ S, image2 f s t := by
  rw [sInter_eq_biInter]
   exact image2_iInter₂_subset_left ..
```
Same below

File: Mathlib/Algebra/Group/Pointwise/Set/Basic.lean (hunk 12 above)
Comment (mans0954) [reply Mathlib/Algebra/Group/Pointwise/Set/Basic.lean]: Done, thanks.

**Review by YaelDillies — APPROVED** [review]
Thanks!

maintainer merge

**Comment by b-mehta** [comment]
Thanks!

bors merge

**Pull request closed without merging**